# Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd
df = pd.read_csv('../data/supermarket_sales.csv')

In [11]:
df.head(3)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total      Date  \
0       Health and beauty       74.69         7  26.1415  548.9715  1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200  3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255  3/3/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [3]:
df.groupby('Branch')['Total'].sum()

Branch
A    106200.3705
B    106197.6720
C    110568.7065
Name: Total, dtype: float64

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [4]:
df.loc[df['Branch'] == df.groupby('Branch')['Total'].sum().idxmax(), 'City'].iloc[0]

'Naypyitaw'

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [5]:
invoices_by_category = pd.pivot_table(df, index='Branch', columns='Product line', values='Invoice ID', aggfunc='count')
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [6]:
invoices_by_category['Electronic accessories'].idxmax()


'A'

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


# Творче завдання

**Гіпотези:**

1. Які типи клієнтів (Member vs Normal) генерують найбільший дохід у кожній філії?
2. Які продуктові лінії мають найбільший прибуток і найбільшу частку продажів?
3. Який відсоток покупок припадає на кожен спосіб оплати (Cash / Ewallet / Credit card)?

### Гіпотеза 1. Які типи клієнтів (Member vs Normal) генерують найбільший дохід у кожній філії?

In [7]:
pd.pivot_table(df, index='Branch', columns='Customer type', values='Total', aggfunc='sum')


Customer type      Member     Normal
Branch                              
A              53637.4755  52562.895
B              53704.6860  52492.986
C              56881.2825  53687.424

In [8]:
df.groupby('Branch')['Total'].sum().to_frame()

Total
Branch             
A       106200.3705
B       106197.6720
C       110568.7065

**Висновки та рекомендації:**
- У всіх філіях члени програми лояльності (Members) приносять більше доходу, ніж звичайні покупці. Це підтверджує ефективність програми лояльності та вказує на потенціал росту через залучення Normal-клієнтів у Member. 
- Рекомендується посилити стимули для вступу в програму та інвестувати більше у персоналізовані пропозиції для Member-клієнтів.

### Гіпотеза 2. Які продуктові лінії мають найбільший прибуток і найбільше продажів?

In [9]:
df.groupby('Product line').agg({'gross income':'sum', 'Invoice ID':'count'}).sort_values(by='gross income', ascending=False)

gross income  Invoice ID
Product line                                    
Food and beverages         2673.5640         174
Sports and travel          2624.8965         166
Electronic accessories     2587.5015         170
Fashion accessories        2585.9950         178
Home and lifestyle         2564.8530         160
Health and beauty          2342.5590         152

**Висновки та рекомендації:**
- Найприбутковіша категорія — Food and Beverages. Має найвищий gross income серед усіх категорій. Кількість продажів також висока (174). Це ключова категорія-лідер, на яку можна робити акцент у маркетингу, акціях та просуванні.

- Категорія Fashion accessories має найбільшу кількість продажів (178), але не є першою по прибутку. Це означає, що товари дешевші, середній чек нижчий, категорія добре продається, але приносить менше прибутку на одну транзакцію.

- Sports and travel та Electronic accessories— високий прибуток (входять в ТОП-3) та стабільні продажі.

- Home and lifestyle та Health and beauty — найслабші категорії. Найнижчий прибуток та найменше продажів у Health and beauty. Це може означати: або низький інтерес в конкретних філія, або невдале просування категорії.

- Загальні рекомендації для мережі: 
    - інвестувати більше у Food and Beverages та Fashion accessories, так, як вони є лідерами по прибутку та продажам;
    - просувати та розвивати категорії Sports and Travel та Electronic Accessories, тому що обидві категорії демонструють стабільний прибуток і продажі;
    - щодо найслабших категорій Home & Lifestyle та Health & Beauty потрібно провести детальний аналіз цих категорій у розрізі філій та переглянути маркетингову стратегію цих категорій.

### Гіпотеза 3. Який відсоток покупок припадає на кожен спосіб оплати (Cash / Ewallet / Credit card)?

In [10]:
df.Payment.value_counts(normalize=True)

Payment
Ewallet        0.345
Cash           0.344
Credit card    0.311
Name: proportion, dtype: float64

**Висновки та рекомендації:**
- Загальна картина показує, що клієнти активно користуються як електронними гаманцями, так і готівкою, а оплата карткою трохи менш поширена.

- Рекомендації:
    - Збільшити кількість кас самообслуговування, оскільки частка безготівкових платежів (Ewallet та Card становить близько 65%);
    - Хоча тренд іде у бік цифрових платежів, все ще варто мати достатньо кас, де можна рораховуватися готівкою, адже частка оплати Cash становить майже 35%.